### DISCUSSION TITLE
Error when running gatsby commands

### DISCUSSION THREAD

**Discussion by ghost** [gatsbyjs/gatsby#33723] [Help] [answered]
### Preliminary Checks

- [X] This issue is not a duplicate. Before opening a new issue, please search existing issues: https://github.com/gatsbyjs/gatsby/issues
- [X] This issue is not a question, feature request, RFC, or anything other than a bug report directly related to Gatsby. Please post those things in GitHub Discussions: https://github.com/gatsbyjs/gatsby/discussions

### Description

When I run the gatsby build or the gatsby develop, errors arrives.
With gatsby build, errors arrive like this:
```bash

 ERROR 

UNHANDLED REJECTION error:0308010C:digital envelope routines::unsupported



  Error: error:0308010C:digital envelope routines::unsupported

  - hash:67 new Hash
    node:internal/crypto/hash:67:19

  - node:crypto:130 Object.createHash
    node:crypto:130:10

  - createHash.js:155 BulkUpdateDecorator.hashFactory
    [01fauna_netlify_gatsby]/[webpack]/lib/util/createHash.js:155:18

  - createHash.js:55 BulkUpdateDecorator.update
    [01fauna_netlify_gatsby]/[webpack]/lib/util/createHash.js:55:51

  - DefinePlugin.js:568 
    [01fauna_netlify_gatsby]/[webpack]/lib/DefinePlugin.js:568:16

  - Array.forEach

  - DefinePlugin.js:564 walkDefinitionsForValues
    [01fauna_netlify_gatsby]/[webpack]/lib/DefinePlugin.js:564:31

  - DefinePlugin.js:591 
    [01fauna_netlify_gatsby]/[webpack]/lib/DefinePlugin.js:591:5


  - Hook.js:14 Hook.CALL_DELEGATE [as _call]
    [01fauna_netlify_gatsby]/[tapable]/lib/Hook.js:14:14

  - Compiler.js:1053 Compiler.newCompilation
    [01fauna_netlify_gatsby]/[webpack]/lib/Compiler.js:1053:26

  - Compiler.js:1097 
    [01fauna_netlify_gatsby]/[webpack]/lib/Compiler.js:1097:29


  - Hook.js:18 Hook.CALL_ASYNC_DELEGATE [as _callAsync]
    [01fauna_netlify_gatsby]/[tapable]/lib/Hook.js:18:14

  - Compiler.js:1092 Compiler.compile
    [01fauna_netlify_gatsby]/[webpack]/lib/Compiler.js:1092:28

  - Compiler.js:517 
    [01fauna_netlify_gatsby]/[webpack]/lib/Compiler.js:517:12

not finished onPreBootstrap - 3.111s
not finished Compiling Gatsby Functions - 3.104s
```

On gatsby develop, errors arrive like this:
```
E:\fauna\01fauna_netlify_gatsby\node_modules\yoga-layout-prebuilt\yoga-layout\build\Release\nbind.js:53
        throw ex;
        ^

Error: error:0308010C:digital envelope routines::unsupported
    at new Hash (node:internal/crypto/hash:67:19)
    at Object.createHash (node:crypto:130:10)
    at BulkUpdateDecorator.hashFactory (E:\fauna\01fauna_netlify_gatsby\node_modules\webpack\lib\util\createHash.js:155:18)
    at BulkUpdateDecorator.update (E:\fauna\01fauna_netlify_gatsby\node_modules\webpack\lib\util\createHash.js:55:51)      
    at E:\fauna\01fauna_netlify_gatsby\node_modules\webpack\lib\DefinePlugin.js:568:16
    at Array.forEach (<anonymous>)
    at walkDefinitionsForValues (E:\fauna\01fauna_netlify_gatsby\node_modules\webpack\lib\DefinePlugin.js:564:31)
    at E:\fauna\01fauna_netlify_gatsby\node_modules\webpack\lib\DefinePlugin.js:591:5
    at Hook.eval (eval at create (E:\fauna\01fauna_netlify_gatsby\node_modules\tapable\lib\HookCodeFactory.js:19:10), <anonymous>:100:1)
    at Hook.CALL_DELEGATE [as _call] (E:\fauna\01fauna_netlify_gatsby\node_modules\tapable\lib\Hook.js:14:14)
    at Compiler.newCompilation (E:\fauna\01fauna_netlify_gatsby\node_modules\webpack\lib\Compiler.js:1053:26)
    at E:\fauna\01fauna_netlify_gatsby\node_modules\webpack\lib\Compiler.js:1097:29
    at Hook.eval [as callAsync] (eval at create (E:\fauna\01fauna_netlify_gatsby\node_modules\tapable\lib\HookCodeFactory.js:33:10), <anonymous>:6:1)
    at Hook.CALL_ASYNC_DELEGATE [as _callAsync] (E:\fauna\01fauna_netlify_gatsby\node_modules\tapable\lib\Hook.js:18:14)
    at Compiler.compile (E:\fauna\01fauna_netlify_gatsby\node_modules\webpack\lib\Compiler.js:1092:28)
    at E:\fauna\01fauna_netlify_gatsby\node_modules\webpack\lib\Watching.js:200:19 {
  opensslErrorStack: [ 'error:03000086:digital envelope routines::initialization error' ],
  library: 'digital envelope routines',
  reason: 'unsupported',
  code: 'ERR_OSSL_EVP_UNSUPPORTED'
}
```

### Reproduction Link

https://github.com/gatsby/gatsby-starter-hello-world

### Steps to Reproduce

1. Run gatsby develop or gatsby build
2. Error will be thrown.

### Expected Result

It should be successful.

### Actual Result

```bash

 ERROR 

UNHANDLED REJECTION error:0308010C:digital envelope routines::unsupported



  Error: error:0308010C:digital envelope routines::unsupported

  - hash:67 new Hash
    node:internal/crypto/hash:67:19

  - node:crypto:130 Object.createHash
    node:crypto:130:10

  - createHash.js:155 BulkUpdateDecorator.hashFactory
    [01fauna_netlify_gatsby]/[webpack]/lib/util/createHash.js:155:18

  - createHash.js:55 BulkUpdateDecorator.update
    [01fauna_netlify_gatsby]/[webpack]/lib/util/createHash.js:55:51

  - DefinePlugin.js:568 
    [01fauna_netlify_gatsby]/[webpack]/lib/DefinePlugin.js:568:16

  - Array.forEach

  - DefinePlugin.js:564 walkDefinitionsForValues
    [01fauna_netlify_gatsby]/[webpack]/lib/DefinePlugin.js:564:31

  - DefinePlugin.js:591 
    [01fauna_netlify_gatsby]/[webpack]/lib/DefinePlugin.js:591:5


  - Hook.js:14 Hook.CALL_DELEGATE [as _call]
    [01fauna_netlify_gatsby]/[tapable]/lib/Hook.js:14:14

  - Compiler.js:1053 Compiler.newCompilation
    [01fauna_netlify_gatsby]/[webpack]/lib/Compiler.js:1053:26

  - Compiler.js:1097 
    [01fauna_netlify_gatsby]/[webpack]/lib/Compiler.js:1097:29


  - Hook.js:18 Hook.CALL_ASYNC_DELEGATE [as _callAsync]
    [01fauna_netlify_gatsby]/[tapable]/lib/Hook.js:18:14

  - Compiler.js:1092 Compiler.compile
    [01fauna_netlify_gatsby]/[webpack]/lib/Compiler.js:1092:28

  - Compiler.js:517 
    [01fauna_netlify_gatsby]/[webpack]/lib/Compiler.js:517:12

not finished onPreBootstrap - 3.111s
not finished Compiling Gatsby Functions - 3.104s
```

### Environment

```shell
System:
    OS: Windows 10 10.0.19042
    CPU: (4) x64 Intel(R) Core(TM) i3-2120 CPU @ 3.30GHz    
  Binaries:
    Node: 17.0.1 - E:\nodejs\node.EXE
    Yarn: 1.22.10 - ~\AppData\Roaming\npm\yarn.CMD
    npm: 8.1.1 - E:\nodejs\npm.CMD
  Languages:
    Python: 3.10.0 - /c/Python310/python
  Browsers:
    Edge: Spartan (44.19041.1266.0), Chromium (95.0.1020.30)
  npmGlobalPackages:
    gatsby-cli: 3.14.0
```

**Comment by LekoArts** [comment]
Hi!

Sorry to hear you're running into an issue. To help us best begin debugging the underlying cause, it is incredibly helpful if you're able to create a [minimal reproduction][reproduction]. This is a simplified example of the issue that makes it clear and obvious what the issue is and how we can begin to debug it.

If you're up for it, we'd very much appreciate if you could [provide a minimal reproduction][reproduction] and we'll be able to take another look.

Thanks for using Gatsby! 💜

[reproduction]: https://gatsby.dev/reproduction

**Comment by ghost** [comment]
Thanks for the nice response. 😊
Here is the issue repository:
https://github.com/SaadFarhanIdress/issue-repo

I previously upgraded from Node 7 to Node 17.
I searched about this error on the internet. I found that I have to downgrade to 14. But It is no more presented on the Node JS website. And also I don't have nvm because it is not supported by windows.

To install dependencies run on the root directory of the repo.
```
# YARN
yarn && cd functions && cd add_message && yarn 
# NPM 
npm i && cd functions && cd add_message && npm i 
```

And then run,
```
gatsby develop
# OR
gatsby build
# OR
netlify dev
```

**Comment by JesHansen** [comment]
This is because you are using Node version 17. 

Workaround: Invoke node manually and instruct it to use the legacy openssl provider like so:

```
node --openssl-legacy-provider node_modules/.bin/gatsby build
node --openssl-legacy-provider node_modules/.bin/gatsby develop

```
More info at https://stackoverflow.com/questions/69665222/node-17-0-1-causes-some-error-digital-envelope-routinesunsupported

**Comment by ghost** [comment]
When I tried, the error appeared, I have installed node_modules installed and gatsby is presented on the .bin folder.
```bash
basedir=$(dirname "$(echo "$0" | sed -e 's,\\,/,g')")
          ^^^^^^^

SyntaxError: missing ) after argument list
    at Object.compileFunction (node:vm:352:18)
    at wrapSafe (node:internal/modules/cjs/loader:1025:15)
    at Module._compile (node:internal/modules/cjs/loader:1059:27)
    at Object.Module._extensions..js (node:internal/modules/cjs/loader:1147:10)
    at Module.load (node:internal/modules/cjs/loader:975:32)
    at Function.Module._load (node:internal/modules/cjs/loader:822:12)
    at Function.executeUserEntryPoint [as runMain] (node:internal/modules/run_main:81:12)
    at node:internal/main/run_main_module:17:47

Node.js v17.0.1
```

**Comment by LekoArts** [comment, marked as answer]
Please try Node.js v16 or Node.js >= 14.15.0

**Reply by ghost** [reply]
I used nvm-windows. And it worked. Sorry for disturbing 😊 .

**Reply by Nantris** [reply]
@LekoArts what about now, from the future, where Node 18 is the new LTS? We have to downgrade to Node 16?

**Comment by bre7** [comment]
Full answer https://stackoverflow.com/a/69665354/2124535

* Use `--openssl-legacy-provider` with Node 17 (`npx cross-env NODE_OPTIONS=--openssl-legacy-provider && npx gatsby build`)
* Downgrade to NodeJS 16

**Reply by obukhov** [reply]
Had same issue on node:18-bullseye
Setting --openssl-legacy-provider didn't help as node reported this parameter to be invalid.

Downgrading to node:16-bullseye worked

**Comment by nikoheikkila** [comment]
Are there any updates on this issue? Node.js 18 is now the LTS version, yet Gatsby cannot be built with that version forcing many sites to work with v16.

**Comment by mockee** [comment]
- node v18.12.1
- Gatsby 5.0.1

```
$ node --openssl-legacy-provider node_modules/.bin/gatsby build
```
And then got an error:

```
UNHANDLED REJECTION Initiated Worker with invalid execArgv flags: --openssl-legacy-provider
  Error: Initiated Worker with invalid execArgv flags: --openssl-legacy-provider
```

**Comment by nikoheikkila** [comment]
I seem to have found something which might help.

I typically use Node.js through the [FNM](https://github.com/Schniz/fnm) version manager or via Homebrew. When I install Node.js 18 from Homebrew using the `brew install node@18` command, all the Gatsby builds work fine in my system.

However, if I install Node.js 18 from FNM using the `fnm install 18` command, I get the same OpenSSL-related error. I suspect the same issue concerns the NVM version manager as well.

So, I dug deeper and discovered that Homebrew-installed Node.js still uses the older OpenSSL version.

You can find the following dependencies by running `node -e 'console.log(process.versions)'` in your shell.

**Homebrew:**

```js
{
  node: '18.12.1',
  v8: '10.2.154.15-node.12',
  uv: '1.44.2',
  zlib: '1.2.11',
  brotli: '1.0.9',
  ares: '1.18.1',
  modules: '108',
  nghttp2: '1.50.0',
  napi: '8',
  llhttp: '6.0.10',
  openssl: '1.1.1s',
  cldr: '41.0',
  icu: '71.1',
  tz: '2022a',
  unicode: '14.0'
}
```

**FNM:**

```js
{
  node: '18.12.1',
  v8: '10.2.154.15-node.12',
  uv: '1.43.0',
  zlib: '1.2.11',
  brotli: '1.0.9',
  ares: '1.18.1',
  modules: '108',
  nghttp2: '1.47.0',
  napi: '8',
  llhttp: '6.0.10',
  openssl: '3.0.7+quic',
  cldr: '41.0',
  icu: '71.1',
  tz: '2022b',
  unicode: '14.0',
  ngtcp2: '0.8.1',
  nghttp3: '0.7.0'
}
```

My suspicion is that Gatsby 5 + Node.js 18 + OpenSSL 1.1 is fine, _but_ Gatsby 5 + Node.js 18 + OpenSSL 3.0 is not. It would make sense, given the workaround flag `--openssl-legacy-provider` does work for some.

How about you, @mockee — can you or anyone else confirm these? Should we also raise a new issue with the Gatsby maintainers?

**Reply by mockee** [reply]
Homebrew (`openssl: '1.1.1s'`) ✅

```js
{
  node: '18.12.1',
  v8: '10.2.154.15-node.12',
  uv: '1.44.2',
  zlib: '1.2.11',
  brotli: '1.0.9',
  ares: '1.18.1',
  modules: '108',
  nghttp2: '1.50.0',
  napi: '8',
  llhttp: '6.0.10',
  openssl: '1.1.1s',
  cldr: '41.0',
  icu: '71.1',
  tz: '2022a',
  unicode: '14.0'
}
```

macOS Installer (`openssl: '3.0.7+quic'`) ❌

```js
{
  node: '18.12.1',
  v8: '10.2.154.15-node.12',
  uv: '1.43.0',
  zlib: '1.2.11',
  brotli: '1.0.9',
  ares: '1.18.1',
  modules: '108',
  nghttp2: '1.47.0',
  napi: '8',
  llhttp: '6.0.10',
  openssl: '3.0.7+quic',
  cldr: '41.0',
  icu: '71.1',
  tz: '2022b',
  unicode: '14.0',
  ngtcp2: '0.8.1',
  nghttp3: '0.7.0'
}
```

**Reply by espipj** [reply]
Using NVM
```

{
  node: '20.10.0',
  acorn: '8.10.0',
  ada: '2.7.2',
  ares: '1.20.1',
  base64: '0.5.0',
  brotli: '1.0.9',
  cjs_module_lexer: '1.2.2',
  cldr: '43.1',
  icu: '73.2',
  llhttp: '8.1.1',
  modules: '115',
  napi: '9',
  nghttp2: '1.57.0',
  nghttp3: '0.7.0',
  ngtcp2: '0.8.1',
  openssl: '3.0.12+quic',
  simdutf: '3.2.18',
  tz: '2023c',
  undici: '5.26.4',
  unicode: '15.0',
  uv: '1.46.0',
  uvwasi: '0.0.19',
  v8: '11.3.244.8-node.25',
  zlib: '1.2.13.1-motley'
}
```

**Comment by MrDesjardins** [comment]
I am having the same error migrating from Gatsby version 3 to 5. Following the two migration documentations ([V3 to V4](https://www.gatsbyjs.com/docs/reference/release-notes/migrating-from-v3-to-v4/) and [V4 to V5](https://www.gatsbyjs.com/docs/reference/release-notes/migrating-from-v4-to-v5/)) I fail to see any mention of that error.

In my case, I am using a Windows machine with WSL. The error appeared after following all these migration steps and executing `npm run build`. 

From a few searches online (like this [Stackoverflow.com](https://stackoverflow.com/questions/69962209/what-is-openssl-legacy-provider-in-node-js-v17)), I suspect it is caused by the version of NodeJS. The accepted answer mentions: `Please try Node.js v16 or Node.js >= 14.15.0` but the [documentation specifies to use a minimum of version 18](https://www.gatsbyjs.com/docs/reference/release-notes/migrating-from-v4-to-v5/#minimal-nodejs-version-1800).

Here is a snippet of my package.json

```
"scripts": {
    "develop": "NODE_OPTIONS='--openssl-legacy-provider' && gatsby develop",
    "debug": "node --openssl-legacy-provider --nolazy node_modules/.bin/gatsby develop --inspect-brk",
    "start": "NODE_OPTIONS='--openssl-legacy-provider' && gatsby develop",
    "build": "NODE_OPTIONS='--openssl-legacy-provider --max_old_space_size=8192' && gatsby build",
    "buildmemory": "NODE_OPTIONS='--max_old_space_size=8192' npm run build",
    "serve": "NODE_OPTIONS='--openssl-legacy-provider' && gatsby serve",
    "clean": "gatsby clean",
    "type-check": "tsc --noEmit",
    "lint": "eslint . --ignore-path .gitignore --ext .ts,.tsx,.js"
  }
```

**Reply by rburgst** [reply]
same happening on mac OSX with volta. 
```
volta run node -e 'console.log(process.versions)'
{
  node: '18.12.1',
  v8: '10.2.154.15-node.12',
  uv: '1.43.0',
  zlib: '1.2.11',
  brotli: '1.0.9',
  ares: '1.18.1',
  modules: '108',
  nghttp2: '1.47.0',
  napi: '8',
  llhttp: '6.0.10',
  openssl: '3.0.7+quic',
  cldr: '41.0',
  icu: '71.1',
  tz: '2022b',
  unicode: '14.0',
  ngtcp2: '0.8.1',
  nghttp3: '0.7.0'
}
```
any guidance from gatsby would be appreciated.

**Reply by rburgst** [reply]
for anyone interested, I did the following to overcome the problem

1. I was using yarn 3.3.2 with node_modules linker -> switched to pnp
2. deleted my yarn.lock file and ran `yarn` again

**Reply by nikoheikkila** [reply]
@rburgst Thank you! Removing `yarn.lock` and running `yarn` again worked for me as well.

**Reply by Nantris** [reply]
This is the only thing that seemed to help but it's not good practice and the upgrade path should really be managed better.

**Comment by Nantris** [comment]
We're on Node 18 and we continue to get the same error. What can we try?

**Comment by bageren** [comment]
Node 18, Gatsby 5, same issue. We would love to to get this solved so we can start using Gatsby 5.

**Reply by Nantris** [reply]
Deleting `yarn.lock` was the only thing that helped me in the end.

**Reply by bageren** [reply]
Unfortunately that didn't work for me :/

**Reply by Nantris** [reply]
Very sorry to hear that. I hope a maintainer might have some advice.

**Reply by bageren** [reply]
Turns out we had the following in our `package.json`:

`"resolutions": {
    "webpack": "5.38.1",
  },`

Removing that solved our issue.

**Comment by EABych** [comment]
Hello guys!
I faced a similar problem, tried the ways of its solution described above, maybe you can help me😔
We use: `"gatsby": "^2.24.63",` and `node:  14.15.4` for our project and it is unfortunately not possible to update the gatsby version right now
last successful build was in May (2 months ago). Yesterday we launched the build - and it failed.
I tried updating the Node version and adding openssl-legacy-provider(from [stackoverflow](https://stackoverflow.com/a/69746387)) , use version 14 of Node, but it didn't work.😔
But I still get an error :
`success Rewriting compilation hashes - 0.006s

 ERROR #98123  WEBPACK

Generating SSR bundle failed

Unexpected token (48:9)

File: node_modules/htmlparser2/lib/esm/index.js:48:9

not finished Building HTML renderer - 18.060s

error Command failed with exit code 1.
`

Can anyone help what should I do?

**Comment by 1-800-jono** [comment]
- Gatsby v5 has been out for a while
- Docs say [Minimal Node.js version 18.0.0](https://www.gatsbyjs.com/docs/reference/release-notes/migrating-from-v4-to-v5/#minimal-nodejs-version-1800)
- Only "solid" solution is to downgrade to node 16

Am I missing something? 😅

**Reply by rdela** [reply]
did you delete yarn.lock https://github.com/gatsbyjs/gatsby/discussions/33723#discussioncomment-4340611 

and any resolutions in package.json https://github.com/gatsbyjs/gatsby/discussions/33723#discussioncomment-6595656

**Comment by espipj** [comment]
When is this planned to be tackled and fixed? This is an issue in all the current Node supported versions: >18. 16 went out of security support 4 months ago... I'm running Gatsby with webpack@5.89.0 and this is still an issue... For more context this webpack issue might help: https://github.com/webpack/webpack/issues/14532

I think this point from @nikoheikkila above is really insightful:
> I seem to have found something which might help.
> 
> I typically use Node.js through the [FNM](https://github.com/Schniz/fnm) version manager or via Homebrew. When I install Node.js 18 from Homebrew using the `brew install node@18` command, all the Gatsby builds work fine in my system.
> 
> However, if I install Node.js 18 from FNM using the `fnm install 18` command, I get the same OpenSSL-related error. I suspect the same issue concerns the NVM version manager as well.
> 
> So, I dug deeper and discovered that Homebrew-installed Node.js still uses the older OpenSSL version.
> 
> You can find the following dependencies by running `node -e 'console.log(process.versions)'` in your shell.
> 
> **Homebrew:**
> 
> ```js
> {
>   node: '18.12.1',
>   v8: '10.2.154.15-node.12',
>   uv: '1.44.2',
>   zlib: '1.2.11',
>   brotli: '1.0.9',
>   ares: '1.18.1',
>   modules: '108',
>   nghttp2: '1.50.0',
>   napi: '8',
>   llhttp: '6.0.10',
>   openssl: '1.1.1s',
>   cldr: '41.0',
>   icu: '71.1',
>   tz: '2022a',
>   unicode: '14.0'
> }
> ```
> 
> **FNM:**
> 
> ```js
> {
>   node: '18.12.1',
>   v8: '10.2.154.15-node.12',
>   uv: '1.43.0',
>   zlib: '1.2.11',
>   brotli: '1.0.9',
>   ares: '1.18.1',
>   modules: '108',
>   nghttp2: '1.47.0',
>   napi: '8',
>   llhttp: '6.0.10',
>   openssl: '3.0.7+quic',
>   cldr: '41.0',
>   icu: '71.1',
>   tz: '2022b',
>   unicode: '14.0',
>   ngtcp2: '0.8.1',
>   nghttp3: '0.7.0'
> }
> ```
> 
> My suspicion is that Gatsby 5 + Node.js 18 + OpenSSL 1.1 is fine, _but_ Gatsby 5 + Node.js 18 + OpenSSL 3.0 is not. It would make sense, given the workaround flag `--openssl-legacy-provider` does work for some.
> 
> How about you, @mockee — can you or anyone else confirm these? Should we also raise a new issue with the Gatsby maintainers?
